tools/garbage_collection/main.go

```diff
@@ -0,0 +1,280 @@
+package main
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"math"
+	"math/rand"
+	"os"
+	"os/exec"
+	"os/user"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"syscall"
+	"time"
+)
+
+// find the glide_XXXX entries in the cwd, matching:
+//   - ownership of the current user
+//   - containing a _GLIDE_LEASE_FILE with mtime greater than 60 minutes
+//   - or if no such file exist, a ctime of more than 10 days
+//   - exclude the dir of the current glidein
+func FindCandidates(excludeDir string) ([]string, float64) {
+
+	var candidates []string
+	startTime := time.Now()
+
+	now := time.Now()
+	currentUser, err := user.Current()
+	uid, err := strconv.Atoi(currentUser.Uid)
+	if err != nil {
+		fmt.Printf("Error determining my own uid: %v", err)
+		return candidates, time.Now().Sub(startTime).Seconds()
+	}
+
+	// Use Glob to find matching glidein dirs
+	dirs, err := filepath.Glob("glide_*")
+	if err != nil {
+		return candidates, time.Now().Sub(startTime).Seconds()
+	}
+
+	// Filter dirs
+	for _, dir := range dirs {
```

Comment:
Could you switch this to be based on [WalkFunc](https://pkg.go.dev/path/filepath#WalkFunc)?

I like `WalkFunc` better because it doesn't require us to completely iterate through the directory -- significantly reduces the load on the filesystem and reduces what's buffered in memory.

Reply:
If we want a random pick, I think we have to build the complete list of candidates first (which means full listing, and stating).

One option would be to relax that requirement, use `WalkFunc`, and select the candidates one by one. However, I suspect this mean the removal happens in lexical order and that could be problematic if we have multiple active glideins.